### PR TITLE
migrate EBNF codegen to DSL v2

### PR DESCRIPTION
**Pull request by OmarTawfik**
leaving the older ebnf codegen in place until `codegen_spec` is migrated as well.

### CODE REVIEW

File: crates/codegen/ebnf/src/builder.rs [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+                Some(Value::new(Self::build_keyword_value(value), None)),
+                DefinitionKind::Sequence,
+            );
+        }
+    }
+
+    fn build_token_item(&mut self, token_item: &TokenItem) {
+        let TokenItem { name, definitions } = token_item;
+
+        self.add_entry(name, true, false);
```
Comment (Xanewok) [inline crates/codegen/ebnf/src/builder.rs:315]: multiple bare boolean arguments are hard to follow when reading multiple of these functions (wait, was the first one terminal or inlined?).

Simplest idea is to add simple two-state enums and use them; other than that, it makes sense to encapsulate that logic in a trait that allows you to create an EBNF ID from the named `XYZItem` variant. This would have the added benefit of encoding the fact, that `add_entry` is only really callable with XYZItem definitions (groups calls, facilitates reasoning), rather than a simple (Identifier, bool, bool) tuple.

Also, since the pattern since to be `add_entry(..);` followed by `self.add_definition();` which only pushes to a vec inside a given map entry, it might make sense to see if we can create a version of `add_entry` that returns the entry and allows you to directly append `Definition`s for simplicity but that's well into nice-to-have area; just thinking out loud.

File: crates/solidity/outputs/spec/build.rs [hunk 2]
```diff
@@ -1,12 +1,88 @@
...
+            writer.buffer.push_str(&format!(
+                "(* {section_index}.{topic_index}. {topic_title}: *)\n\n",
+                section_index = section_index + 1,
+                topic_index = topic_index + 1,
+                topic_title = topic.title,
+            ));
+
+            for item in &topic.items {
+                ebnf_model.serialize(item.name(), &mut writer);
+                writer.buffer.push('\n');
```
Comment (Xanewok) [inline crates/solidity/outputs/spec/build.rs:52]: We used `end_line` in serializer - maybe it makes sense to do this here, as well?

File: crates/solidity/outputs/spec/build.rs [hunk 3]
```diff
@@ -1,12 +1,88 @@
...
+            section_title = section.title,
+        ));
+
+        for (topic_index, topic) in section.topics.iter().enumerate() {
+            writer.buffer.push_str(&format!(
+                "(* {section_index}.{topic_index}. {topic_title}: *)\n\n",
+                section_index = section_index + 1,
+                topic_index = topic_index + 1,
+                topic_title = topic.title,
+            ));
```
Comment (Xanewok) [inline crates/solidity/outputs/spec/build.rs:48]: We already have access to `write_comment` - I think it'd be a bit cleaner about the intent if we use those, instead of the manual `push_str`

File: crates/solidity/outputs/spec/build.rs [hunk 4]
```diff
@@ -1,12 +1,88 @@
...
+        buffer: String::new(),
+    };
+    let ebnf_model = EbnfModel::build(language);
+
+    for (section_index, section) in language.sections.iter().enumerate() {
+        writer.buffer.push_str(&format!(
+            "(*\n * {section_index}. {section_title}:\n *)\n\n",
+            section_index = section_index + 1,
+            section_title = section.title,
+        ));
```
Comment (Xanewok) [inline crates/solidity/outputs/spec/build.rs:40]: Same as below; it'd be nice to only go through the `EbnfWriter` here (and I don't think multi-line gives us much, if we're worried about abstraction leakage and having to include ` *` prefix explicitly)

File: crates/codegen/ebnf/src/serializer.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            self.serialize_expr(child);
+        }
+    }
+
+    fn serialize_comment(&mut self, value: &String) {
+        self.writer.write_comment(format!("(* {value} *)"));
+    }
+
+    fn serialize_identifier(&mut self, name: &str) {
+        let value = &self.model.entries[name].ebnf_id;
```
Comment (Xanewok) [inline crates/codegen/ebnf/src/serializer.rs:157]: We don't emit the same panic message as above, in `fn serialize`. Is this intended?

**Review by Xanewok — CHANGES_REQUESTED** [review]
Looks good! I have some readability nitpicks in builder.rs that I'd like to see addressed, if you don't mind :pray:

File: crates/codegen/ebnf/src/builder.rs (hunk 1 above)
Comment (OmarTawfik) [reply crates/codegen/ebnf/src/builder.rs:315]: I agree as well if this was a publicly used method, but having a trait and 9-10 implementations for a private helper used within one type is too much work/too much complex for the benefit it adds IMO.

Also, `add_entry()` and `add_definition()` are not always called together, as many items add one entry and multiple definitions. So I think the uniformity here is worth it.

File: crates/codegen/ebnf/src/builder.rs [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+            Some(VersionSpecifier::Never) => Some("Never reserved".to_string()),
+            Some(VersionSpecifier::From { from }) => Some(format!("Reserved in {from}")),
+            Some(VersionSpecifier::Till { till }) => Some(format!("Reserved until {till}")),
+            Some(VersionSpecifier::Range { from, till }) => {
+                Some(format!("Reserved from {from} until {till}"))
+            }
+        }
+    }
+
+    fn build_enabled(enabled: &Option<VersionSpecifier>) -> Option<String> {
```
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply crates/codegen/ebnf/src/builder.rs:371]: Good point. Will rename them.

File: crates/codegen/ebnf/src/serializer.rs (hunk 5 above)
Comment (OmarTawfik) [reply crates/codegen/ebnf/src/serializer.rs:157]: The above took its input from external callers, so more likely to fail. But it makes sense to show a message in both. Will modify.

File: crates/codegen/ebnf/src/builder.rs [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+        name: &Identifier,
+        leading_comments: impl IntoIterator<Item = String>,
+        values: impl IntoIterator<Item = Value>,
+        kind: DefinitionKind,
+    ) {
+        let definition = Definition::new(
+            leading_comments.into_iter().collect(),
+            values.into_iter().collect(),
+            kind,
+        );
```
[the remark this answers is not in the document]
Comment (Xanewok) [reply crates/codegen/ebnf/src/builder.rs:74]: Don't mean to outline this function; it's just that using bare struct construction syntax is simpler and tells that nothing else is going on: 
```rust
let definition = Definition {
  leading_comments: ..,
  value: ..,
  kind,
};
```

Here, `new` only hides what's going on and introduces reliance on the order rather than being explicit. This is not important and is style but wanted to note that.

File: crates/solidity/outputs/spec/build.rs (hunk 2 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/spec/build.rs:52]: I suggest keeping `start_line()`/`end_line()` pairs in sync, and related to serializer use cases.
The newline here is for `SimpleEbnfWriter`, and won't be needed for other implementations, like the HTML one that is used in the spec.

File: crates/solidity/outputs/spec/build.rs (hunk 3 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/spec/build.rs:48]: Sounds good. Will do.

File: crates/solidity/outputs/spec/build.rs (hunk 4 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/spec/build.rs:40]: Sounds good. Will do.

Comment (OmarTawfik) [reply crates/solidity/outputs/spec/build.rs:40]: This writer is solely for writing `.ebnf` files. The abstraction is here to allow us to write other formats, like HTML (syntax higlighting in the rendered spec).

File: crates/codegen/ebnf/src/builder.rs (hunk 1 above)
Comment (Xanewok) [reply crates/codegen/ebnf/src/builder.rs:315]: I understand that the above might be overengineering. In which case, introducing trivial enums like `Terminal::Yes`, `Inline::No` would be a great and simple alternative :+1: 

Reading
```rust
self.add_entry(name, Terminal::Yes, Inlined::No)
```
rather than
```rust
self.add_entry(name, true, false)
```

is a lot more readable and is easier to understand what's going on, I'm sure you agree as well :crossed_fingers:

File: crates/solidity/outputs/spec/build.rs (hunk 3 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/spec/build.rs:48]: This writer is solely for writing `.ebnf` files. The abstraction is here to allow us to write other formats, like HTML (syntax higlighting in the rendered spec).

File: crates/codegen/ebnf/src/builder.rs (hunk 1 above)
Comment (OmarTawfik) [reply crates/codegen/ebnf/src/builder.rs:315]: Sure

File: crates/solidity/outputs/spec/build.rs (hunk 3 above)
Comment (Xanewok) [reply crates/solidity/outputs/spec/build.rs:48]: Agreed, it's just that since we have the simple API ready and exposed, it'd probably be more declarative to say `writer.write_comment(format!(...))` rather than trying to manipulate the buffer directly, reader has to parse the string themselves to see it's a comment etc.

File: crates/codegen/ebnf/src/serializer.rs (hunk 5 above)
Comment (Xanewok) [reply crates/codegen/ebnf/src/serializer.rs]: This makes sense and so we didn't need to change that; just wanted to ask to know if I'm missing something. Sorry for the churn!

File: crates/solidity/outputs/spec/build.rs (hunk 3 above)
Comment (Xanewok) [reply crates/solidity/outputs/spec/build.rs:48]: If the .ebnf file is just to make sure everything is serialized properly and won't be used after the migrate the spec gen to use it, then we probably shouldn't change that and can remove it once we migrate the spec-gen.

Comment (OmarTawfik) [reply crates/solidity/outputs/spec/build.rs:48]: Sure.

File: crates/codegen/ebnf/src/builder.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply crates/codegen/ebnf/src/builder.rs:74]: I agree, but I think that is more important in case if was something like `codegen_language_definition::model::*` where it is used everywhere, and being able to grep/understand it across the entire codebase is important, and also it needs to survive renaming/refactoring over time. But in this case, for these internal types that are only instantiated (repeatedly) in one file, the brevity actually helps make the surrounding code more readable IMO, so that it can focus on the logic, rather than the verbose field/value pairs everywhere. I think this benefit (given how frequently these initializations happen here) outweighs that.

[the remark this answers is not in the document]
Comment (Xanewok) [reply crates/codegen/ebnf/src/builder.rs:74]: Agreed, I'm not blocking on that and I mentioned that it's ultimately not important and mostly a style issue. I noticed the addition of `derive-new` so wanted to mention that sometimes it's helpful but sometimes it's actually worse.

At the cost of stutter/repeating in struct init syntax, we know this has no post-process logic, we know it's a POD, can look up fields in IDE on hover etc.

Alas, I may have donned my Rust Evangelism Strikeforce hat and didn't mean to make this into an actual issue, excuse me :see_no_evil:

File: crates/codegen/ebnf/src/builder.rs [hunk 8]
```diff
@@ -430,3 +433,13 @@ impl Builder {
         Expression::new_reference(reference.to_string())
     }
 }
+
+enum Terminal {
```
Comment (Xanewok) [inline crates/codegen/ebnf/src/builder.rs:437]: These can be easily `Copy` to avoid `&` everywhere

File: crates/solidity/outputs/spec/build.rs [hunk 9]
```diff
@@ -33,14 +33,14 @@ fn generate_language_ebnf(
...
-        writer.buffer.push_str(&format!(
+        writer.write_comment(format!(
             "(*\n * {section_index}. {section_title}:\n *)\n\n",
             section_index = section_index + 1,
             section_title = section.title,
         ));
 
         for (topic_index, topic) in section.topics.iter().enumerate() {
-            writer.buffer.push_str(&format!(
+            writer.write_comment(format!(
```
Comment (Xanewok) [inline crates/solidity/outputs/spec/build.rs:43]: Oh dear, okay, I'm dumb; I didn't put the two and two together here and was ultimately confused by what `SimpleEbnfWriter` does here - it really is just appending to the buffer. In which case, I guess, it's fine to keep the buffer manipulation but why do we need the `EbnfWriter` trait again? Is this going to be used for the later pass, when we rewrite the spec-gen? If so, maybe we could add a comment that it's only used temporarily?

**Review by Xanewok — APPROVED** [review]
I was confused by the `SimpleEbnfWriter` and didn't mean to block you on minor stuff. I'm approving not to induce churn - feel free to either re-add it to the queue or make the helper enums Copy to reduce some of the sigils in call-sites.

File: crates/solidity/outputs/spec/build.rs (hunk 9 above)
Comment (Xanewok) [reply crates/solidity/outputs/spec/build.rs:43]: > The spec will use the other implementation to write actual <span> and <a> tags that are used for syntax highlighting and cross linking.

I missed that, sorry. Feel free to add a comment or not; I'm assuming we will work on this shortly, anyway.

File: crates/codegen/ebnf/src/builder.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (OmarTawfik) [reply crates/codegen/ebnf/src/builder.rs:74]: 🦀

File: crates/codegen/ebnf/src/builder.rs (hunk 8 above)
Comment (OmarTawfik) [reply crates/codegen/ebnf/src/builder.rs:437]: Good point.

File: crates/solidity/outputs/spec/build.rs (hunk 9 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/spec/build.rs:43]: >  Is this going to be used for the later pass, when we rewrite the spec-gen? If so, maybe we could add a comment that it's only used temporarily?

Based on this, I will actually create a general `PlainEbnfWriter` and add comments to it, as we will want to reuse it to write doc comments in kinds.rs.jinja2, similar to what we used to have, but this will probably have to wait for #638 to complete first.
